src/twisted/conch/ssh/session.py

```diff
@@ -180,16 +180,16 @@ def extReceived(self, dataType, data):
             log.warn("Weird extended data: {dataType}", dataType=dataType)
 
     def eofReceived(self):
+        if self.client:
```

Comment:
Yeah, arranging for backward compatibility here is utterly miserable.  We ran into this pretty hard when adding environment variable passing support as well.

`self.client` isn't just a subsystem thing - it's also used for communicating with shells or commands.  But yes, I agree that this is all a mess.  A further complication that I'm not sure you've considered is that environment variable passing requires creating an `ISession` adapter (`self.session`) before we're in a position to create an `SSHSessionProcessProtocol` (`self.client`), and the resulting `ISession` adapter must be kept around since otherwise the environment variable request would have no effect.

One approach that might work in a backward-compatible way would be to leave `SSHSession.client` and `SSHSession.session` as-is (unfortunate though that layout is), but move the session handling in `SSHSession.eofReceived` and `SSHSession.closed` into `SSHSessionProcessProtocol`; that way at least the protocol controls all the cleanup.  However, for that to work properly we'd need to create an unconnected `SSHSessionProcessProtocol` a bit earlier in order to handle the case where an environment variable request has been sent but a shell/exec/subsystem request hasn't yet been sent, since currently `self.session` is set in that case but `self.client` isn't.

I think that could be done without a total backward-incompatible rewrite.  What do you think?  (But it feels like something that should be a separate PR.)

Reply:
> A further complication that I'm not sure you've considered is that environment variable passing requires creating an ISession adapter (self.session) before we're in a position to create an SSHSessionProcessProtocol (self.client), and the resulting ISession adapter must be kept around since otherwise the environment variable request would have no effect.

Yes. I remember reviewing this. Thanks again for that fix :)


-----------

> I think that could be done without a total backward-incompatible rewrite. What do you think? (But it feels like something that should be a separate PR.)

yes. A separate review.
I am happy to review such a PR :)

I will not have time to work on that refactoring anytime soon :(